pgx/src/datum/range.rs

```diff
@@ -9,56 +9,281 @@ Use of this source code is governed by the MIT license that can be found in the
+    /// let _ = Range::<i32>::new(1, None); // `(1..)`
+    /// let _ = Range::<i32>::new(None, RangeBound::Exclusive(10)); // `(..10)`
+    /// let _ = Range::<i32>::new(1, RangeBound::Exclusive(10)); // (`1..10)`
+    /// let _ = Range::<i32>::new(None, None); // `(..)`
+    /// let _ = Range::<i32>::new(RangeBound::Infinite, RangeBound::Infinite); // `(..)`
+    #[inline]
+    pub fn new<L, U>(lower: L, upper: U) -> Self
+    where
+        L: Into<RangeBound<T>>,
+        U: Into<RangeBound<T>>,
+    {
+        Self { inner: Some((lower.into(), upper.into())) }
+    }
+
+    /// Builds an "empty" range
+    ///
+    /// Unlike Rust ranges (from `std::ops::`), Postgres ranges can be empty, meaning they don't
+    /// represent any range of values.
+    #[inline]
+    pub fn empty() -> Self {
+        Self { inner: None }
+    }
+
+    /// Builds an "infinite" range.  This is equivalent to Rust's [`std::ops::RangeFull`] (`(..)`).
+    #[inline]
+    pub fn infinite() -> Self {
+        Self::new(RangeBound::Infinite, RangeBound::Infinite)
+    }
+
+    /// Returns the lower [`RangeBound`]
+    #[inline]
+    pub fn lower(&self) -> Option<&RangeBound<T>> {
+        match &self.inner {
+            Some((l, _)) => Some(l),
+            None => None,
+        }
+    }
+
+    /// Returns the upper [`RangeBound`]
+    #[inline]
+    pub fn upper(&self) -> Option<&RangeBound<T>> {
+        match &self.inner {
+            Some((_, u)) => Some(u),
+            None => None,
         }
     }
+
+    /// Returns 'true' if the range is "empty".
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_none()
+    }
+
+    /// Returns `true` if the range is "infinite".  This is equivalent to Rust's [`std::ops::RangeFull`] (`(..)`)
+    #[inline]
+    pub fn is_infinite(&self) -> bool {
+        match (self.lower(), self.upper()) {
+            (Some(RangeBound::Infinite), Some(RangeBound::Infinite)) => true,
+            _ => false,
+        }
+    }
+
+    /// Consumes `self` and returns the internal representation, which can be easily mapped or
+    /// unwrapped.
+    ///
+    /// A return value of [`Option::None`] indicates that this range represents the "empty" range.
+    #[inline]
+    pub fn into_inner(self) -> Option<(RangeBound<T>, RangeBound<T>)> {
+        self.inner
+    }
+
+    /// Takes the bounds out of this [`Range`] and converts self to represent the "empty" range.
+    ///
+    /// A return value of [`Option::None`] indicates that this range already represents the "empty" range.
+    #[inline]
+    pub fn take(&mut self) -> Option<(RangeBound<T>, RangeBound<T>)> {
+        self.inner.take()
+    }
+
+    /// Consumes `self` and returns the contained [`RangeBound`]s.
+    ///
+    /// # Panics
+    ///
+    /// This function panics if this [`Range`] represents an "empty" range.
+    #[inline]
+    #[track_caller]
+    pub fn unwrap(self) -> (RangeBound<T>, RangeBound<T>) {
+        self.inner.unwrap()
+    }
```

Comment:
This feels like a dubious function to expose since `into_inner` and `take` already exist, unlike `replace` where we need to adjust for the fact that `Option::replace` isn't right for that use.

Reply:
I argued with myself over this one.  Happy to remove it.

Reply:
removed.